caffe2/serialize/intermediate_model.h

```diff
@@ -0,0 +1,600 @@
+  // getters
+  const std::string& name() const {
+    return name_;
+  }
+
+  const std::string& producerName() const {
+    return producerName_;
+  }
+
+  const std::string& producerVersion() const {
+    return producerVersion_;
+  }
+
+  const IntermediateModule& mainModule() const {
+    return mainModule_;
+  }
+
+  int64_t protoVersion() const {
+    return protoVersion_;
+  }
+
+  // setters, most for test purposes
+  void setName(const std::string& name) {
+    name_ = name;
+  }
+
+  void setProducerName(const std::string& producer_name) {
+    producerName_ = producer_name;
+  }
+
+  void setProducerVersion(const std::string& producer_version) {
+    producerVersion_ = producer_version;
+  }
+
+  void setProtoVersion(int64_t proto_version) {
+    protoVersion_ = proto_version;
+  }
+
+  IntermediateModule* mutableMainModule() {
+    return &mainModule_;
+  }
+
+ private:
+  std::string name_;
+  std::string producerName_;
+  std::string producerVersion_;
+  int64_t protoVersion_;
+  IntermediateModule mainModule_;
+
+};
+
+// serialize an IntermediateModel through a PyTorchFileWriter
+// we always put the model data at the end, so when serializing
+// model, the we assume the record_id in imodel is already updated
+void serializeIntermediateModel(IntermediateModel* imodel,
+    torch::jit::PyTorchFileWriter* writer) {
+  std::unordered_map<void*, uint64_t> data_id;
+  std::stack<IntermediateModule*> imodules;
+  imodules.push(imodel->mutableMainModule());
+  while (!imodules.empty()) {
+    IntermediateModule* m = imodules.top();
+    imodules.pop();
+    auto* params = m->mutableParameters();
+    for (int i = 0; i < params->size(); ++i) {
+      std::shared_ptr<SharedData> dataptr = params->at(i).mutableTensor()->data();
+      void* data = dataptr->rawData();
+      size_t size = dataptr->size();
+      auto it = data_id.find(data);
+      if (it != data_id.end()) {
+        dataptr->setRecordId(it->second);
+      } else {
+        uint64_t id = writer->writeRecord(data, size);
+        dataptr->setRecordId(id);
+        data_id[data] = id;
+      }
+    }
+
+    auto* subms = m->mutableSubmodules();
+    for (int i = 0; i < subms->size(); ++i) {
+      imodules.push(&subms->at(i));
+    }
+  }
+
+  torch::ModelDef model_def;
+  imodel->dump(&model_def);
+  size_t proto_size = model_def.ByteSizeLong();
+  void* buffer = malloc(proto_size);
+  model_def.SerializeToArray(buffer, proto_size);
+  writer->writeRecord(buffer, proto_size);
+  free(buffer);
+}
+
+// serialize an IntermediateModel to a given file
+void serializeIntermediateModel(IntermediateModel* imodel, const std::string& filename) {
+  torch::jit::PyTorchFileWriter writer(filename);
+  serializeIntermediateModel(imodel, &writer);
+  writer.writeEndOfFile();
+}
+
+// deserialize an IntermediateModel through a reader,
+// serialize tensors' data first, and maintain the mappint from
+// record id to tensor data
+void deserializeIntermediateModel(IntermediateModel* imodel,
```

Comment:
We took great care writing the original deserializer so that it would not require you to read through all the tensor entries and their memory before reading the model so that loading the model can be fast, and that tensor data could be loaded later by mmapping the data sections of the file. It seems like this approach doesn't support that because it reads through the sections eagerly. Can we just load a tensor when we find it in the model? If the file  was mmap it would be possible to initialize the Tensor without ever reading the Tensor records.

Reply:
Yes, I will add a lazy mode in deserialize, which only load the model metadata, and we can feed the data to the tensor later. Right now we keep the model as the last record, and corresponding functionality in the FileReader.